napari/layers/_tests/test_source.py

```diff
@@ -0,0 +1,35 @@
+from napari.layers import Points
+from napari.layers._source import Source, current_source, layer_source
+
+
+def test_layer_source():
+    """Test basic layer source assignment mechanism"""
+    with layer_source(path='some_path', reader_plugin='builtins'):
+        points = Points()
+
+    assert points.source == Source(path='some_path', reader_plugin='builtins')
+
+
+def test_source_context():
```

Comment:
Yup, test feels super nice. =)

I think in the future we might want to have the concept of nested sources (ie nested context nests, instead of overriding), but we can see how people start using this before getting too fancy.